GitUI/CommandsDialogs/SettingsDialog/Pages/RevisionLinksExtractor.cs

```diff
@@ -0,0 +1,202 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Text.RegularExpressions;
+using System.Threading.Tasks;
+using GitCommands.ExternalLinks;
+using ResourceManager;
+
+namespace GitUI.CommandsDialogs.SettingsDialog.Pages
+{
+    public enum CloudProviderKind
+    {
+        None = 0,
+        GitHub,
+        AzureDevOps,
+    }
+
+    public interface ICloudProviderExternalLinkDefinitionExtractor
+    {
+        bool IsValidRemoteUrl(string remoteUrl);
+        IList<ExternalLinkDefinition> GetDefinitions(string remoteUrl);
+    }
+
+    public abstract class ExternalLinkDefinitionExtractor : ICloudProviderExternalLinkDefinitionExtractor
+    {
+        protected readonly TranslationString _linkCodeToString = new TranslationString("{0} - Code");
+        protected readonly TranslationString _linkIssuesToString = new TranslationString("{0} - Issues");
+        protected readonly TranslationString _linkPullRequestsToString = new TranslationString("{0} - Pull Requests");
+        protected readonly TranslationString _linkViewCommitToString = new TranslationString("View commit in {0}");
+        protected readonly TranslationString _linkViewProjectToString = new TranslationString("View project in {0}");
+
+        protected Match MatchRegExes(string remoteUrl, string[] regExs)
+        {
+            Match m = null;
+            foreach (var regex in regExs)
+            {
+                m = Regex.Match(remoteUrl, regex);
+                if (m.Success)
+                {
+                    break;
+                }
+            }
+
+            return m;
+        }
+
+        public abstract bool IsValidRemoteUrl(string remoteUrl);
+
+        public abstract IList<ExternalLinkDefinition> GetDefinitions(string remoteUrl);
+    }
+
+    public sealed class AzureDevopsExternalLinkDefinitionExtractor : ExternalLinkDefinitionExtractor
+    {
+        public static readonly string VstsHttpsRemoteRegex = @"^https:\/\/(?<owner>[^.]*)\.visualstudio\.com\/(?<project>[^\/]*)\/_git\/(?<repo>.*)$";
```

Comment:
Non interface members (i.e. those not implementing the interface) should not be public